[NicholasCWK] iP

NicholasCWK opened the pull request:
# DukePro
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2425s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :+1:
And it is **FREE**!

Features:

- [x]  Managing tasks
- [ ]  Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```java
public static void main(String[] args) {
        new Fred().run();
    }
```

NicholasCWK committed:
Rename Duke.java to Echo.java

NicholasCWK committed:
Add greeting and farewell

NicholasCWK committed:
Rename Echo.java to Fred.java

NicholasCWK committed:
Change name string from Echo to Fred

NicholasCWK committed:
add greet and farewell methods

NicholasCWK committed:
add exit and echo methods

NicholasCWK committed:
add addToTaskList and printTaskList methods

NicholasCWK committed:
add Task.java

NicholasCWK committed:
add description and isDone fields, getStatusIcon, markAsDone, markAsNotDone methods

NicholasCWK committed:
Change printTaskList printing format and add getDescription

NicholasCWK committed:
Revert "Change printTaskList printing format and add getDescription"

This reverts commit 5085ddda6812e3fe8bb10afb6bb77bcce98615b6.

NicholasCWK committed:
add getDescription

NicholasCWK committed:
modify addToTaskList and printTaskList

NicholasCWK committed:
add getInput method

NicholasCWK committed:
add markTaskAsDone and markTaskAsNotDone methods

NicholasCWK committed:
add Deadline.java, Event.java and ToDo.java

NicholasCWK committed:
Remove getDescription method and modify toString method

NicholasCWK committed:
Add constructor and modify toString method

NicholasCWK committed:
Add field and constructor, modify toString method

NicholasCWK committed:
add fields and constructor, modify toString method

NicholasCWK committed:
modify toString methods of Deadline.java and Event.java

NicholasCWK committed:
modify toString method of Deadline.java, Event.java and Todo.java

NicholasCWK committed:
modify getInput method to detect todo, deadline and event tasks, modify addToTaskList to add todo, deadline and event tasks

NicholasCWK committed:
add test case to expected.txt and input

NicholasCWK committed:
modify runtest.sh to feed commands from input.txt to Fred.java

NicholasCWK committed:
add FredException.java

NicholasCWK committed:
add constructor

NicholasCWK committed:
add EmptyTaskDescriptionException.java

NicholasCWK committed:
add EmptyInputException.java and UnknownCommandException.java

NicholasCWK committed:
add InvalidTaskNumberException.java

NicholasCWK committed:
modify determineAction and addToTaskList methods

NicholasCWK committed:
Add InvalidTaskNumberException.java

NicholasCWK committed:
modify markTaskAsDone and markTasksAsNotDone methods

NicholasCWK committed:
Modify Fred.java to use ArrayList for taskList

NicholasCWK committed:
add deleteTaskFromTaskList

NicholasCWK committed:
Shift exception handling to determineAction method and add say method

NicholasCWK committed:
Change message

NicholasCWK committed:
Fred.java: Add getDataFile method

NicholasCWK committed:
Fred.java: Add loadFromDataFile method

NicholasCWK committed:
Fred.java: Add createTask method

NicholasCWK committed:
Task.java: Add getDataFormat method

NicholasCWK committed:
Event.java: Add getDataFormat method

NicholasCWK committed:
Deadline.java: Add getDataFormat method

NicholasCWK committed:
Todo.java: Add getDataFormat method

NicholasCWK committed:
Fred.java: add rewriteDataFile and appendToDataFile methods

NicholasCWK committed:
Todo.java: Rename Todo constructor

NicholasCWK committed:
Fred.java: Change createTask method

NicholasCWK committed:
Todo.java: Change Todo constructor

NicholasCWK committed:
Fred.java: Change createTask method

NicholasCWK committed:
Merge branch 'branch-Level-7'

NicholasCWK committed:
Fred.java: Change createTask method and loadDataFromFile method

NicholasCWK committed:
Deadline.java: Change fields to store localDateTime

NicholasCWK committed:
Event.java: Change fields to store localDateTime

NicholasCWK committed:
Merge branch 'branch-Level-8'

NicholasCWK committed:
Ui.java: Add say, greet, sayFarewell, printTaskList methods

NicholasCWK committed:
Fred.java: Remove say, greet, sayFarewell, printTaskList methods

NicholasCWK committed:
Parser.java: Add parseInput method

NicholasCWK committed:
Ui.java: Add say, greet, sayFarewell, printTaskList, getInput methods

NicholasCWK committed:
Storage.java: Add getDataFile, getTasksFromDataFile, deleteFromDataFile, appendToDataFile methods

NicholasCWK committed:
TaskList.java: Add createTask, addToTaskList, markTaskAsDone, markTaskAsNotDone, deleteFromTaskList, loadTasksFromDataFile, getTaskListSize, getTaskList methods

NicholasCWK committed:
Fred.java: Refactor code to extract methods

NicholasCWK committed:
Merge branch 'A-MoreOOP'

NicholasCWK committed:
Add classes to fred package

NicholasCWK committed:
Merge branch 'A-Packages'

NicholasCWK committed:
Add gradle support

NicholasCWK committed:
Add gradle support

NicholasCWK committed:
Merge branch 'add-gradle-support'

NicholasCWK committed:
Add ParserTest.java, TodoTest.java

NicholasCWK committed:
Merge branch 'A-JUnit'

* A-JUnit:
  Add ParserTest.java, TodoTest.java

NicholasCWK committed:
Change class name

NicholasCWK committed:
Add META-INF

NicholasCWK committed:
Merge branch 'A-Jar'

* A-Jar:
  Add META-INF
  Change class name

NicholasCWK committed:
Add javadocs

NicholasCWK committed:
Parser.java: Change parseInput method to detect find command

NicholasCWK committed:
TaskList.java: Add findTasksInTaskList method

NicholasCWK committed:
Task.java: Add checkForKeyword method

NicholasCWK committed:
Ui.java: Add printTasksWithKeyword method

NicholasCWK committed:
Fred.java: Change executeAction method to execute find command

NicholasCWK committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Fred.java: Change executeAction method to execute find command
  Ui.java: Add printTasksWithKeyword method
  Task.java: Add checkForKeyword method
  TaskList.java: Add findTasksInTaskList method
  Parser.java: Change parseInput method to detect find command

NicholasCWK committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add javadocs

# Conflicts:
#	src/main/java/fred/Task.java

NicholasCWK committed:
Fred.java: Change indentation to adhere to coding standard

NicholasCWK committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Fred.java: Change indentation to adhere to coding standard

# Conflicts:
#	src/main/java/fred/Fred.java

NicholasCWK committed:
Fred.java: Change executeAction method to execute find command

triishaa commented on src/main/java/fred/Deadline.java:
```diff
@@ -0,0 +1,46 @@
+     * @return A string in the format "D | status | description | due date".
+     */
+    @Override
+    public String getDataFormat() {
```
the naming sounds a bit ambiguous? what do you mean by getDataFormat? shouldnt it be something like printDataFormatted? or printDataAfterFormatting?

triishaa commented on src/main/java/fred/Event.java:
```diff
@@ -0,0 +1,49 @@
+     * @return A string in the format "E | status | description | start time-end time".
+     */
+    @Override
+    public String getDataFormat() {
```
same as above? the naming sounds a bit not-explanatory

triishaa commented on src/main/java/fred/Fred.java:
```diff
@@ -0,0 +1,121 @@
+ */
+public class Fred {
+    Storage storage;
+    TaskList tasks;
```
good following of the coding convention here.

triishaa commented on src/main/java/fred/Fred.java:
```diff
@@ -0,0 +1,121 @@
+     * Runs the Fred application. This method loads tasks from the storage, greets the user,
+     * and enters a loop to process user inputs and execute corresponding actions.
+     */
+    void run() {
```
good use of naming convention here. Clearly understood what the method is used for. 🎉

triishaa commented on src/main/java/fred/Todo.java:
```diff
@@ -0,0 +1,37 @@
+     * @return A string in the format "T | status | description".
+     */
+    @Override
+    public String getDataFormat() {
```
same as above, is the naming explaining wha the function does?

triishaa commented on src/main/java/fred/Ui.java:
```diff
@@ -0,0 +1,81 @@
+     *
+     * @param message The message to be displayed to the user.
+     */
+    void say(String message) {
```
naming is a bit ambiguous, say what? maybe something like sayCustom? cause you used sayFarewell for farewell messge? so maybe?

KengHian commented on src/main/java/fred/TaskList.java:
```diff
@@ -0,0 +1,142 @@
+     * @return The task that was deleted.
+     * @throws FredException If the task number is invalid.
+     */
+    Task deleteFromTaskList(int taskNumber) throws FredException {
```
I think perhaps a better name for this method would be `deleteTaskFromList`?

KengHian commented on src/main/java/fred/TaskList.java:
```diff
@@ -0,0 +1,142 @@
+ * and load tasks from an external data source.
+ */
+public class TaskList {
+    ArrayList<Task> taskList = new ArrayList<>();
```
Perhaps a better name would just be `tasks`? Just my opinion though 😅.

KengHian commented on src/main/java/fred/TaskList.java:
```diff
@@ -0,0 +1,142 @@
+     *
+     * @param task The task to be added to the list.
+     */
+    void addToTaskList(Task task) {
```
Similar to the comment below, where I personally think `addTaskToList` is a clearer name.

KengHian commented on src/main/java/fred/Fred.java:
```diff
@@ -0,0 +1,121 @@
+    /**
+     * Exits the Fred application by terminating the program.
+     */
+    void exit() {
```
Just curious, may I ask why are some of the functions in this class without access modifier?

KengHian commented on src/main/java/fred/Parser.java:
```diff
@@ -0,0 +1,68 @@
+                return new String[]{"printTaskList"};
+            } else if (inputParts[0].equals("mark") || inputParts[0].equals("unmark") || inputParts[0].equals("delete")) {
+                throw new InvalidTaskNumberException();
+            } else if (inputParts[0].equals("todo") || inputParts[0].equals("deadline") || inputParts[0].equals("event") || inputParts[0].equals("find")) {
```
I think this line has exceed 120 chars, and you could place some line break to improve readability.

KengHian reviewed (commented):
Overall I found your code to be clear and well structure. While I am suppose to focus on code quality, I found your code standard very consistent, one thing to take note is to use line break on long line (120 chars) of code for readability. 

As for your code quality for naming, there also is not too many issues I can point out, most variable/method names are clear at describing what they do. Therefore what I commented on are honestly just nitpick. 

Lastly, good job on the detailed JavaDoc and the implementation of the various exception!

NicholasCWK committed:
Add GUI

NicholasCWK committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Add GUI

NicholasCWK committed:
Group exception files and task files into respective folders

NicholasCWK committed:
Add Command and TaskType enums

NicholasCWK committed:
Add Action class

NicholasCWK committed:
Refactor Fred, Parser and TaskList class

The parseInput method returns a string array with magic strings (e.g.
“printTaskList”, “sayFarewell” etc.).

The executeAction method receives the string array and executes the
respective actions based on these magic strings.

Using named constants will eliminate the use of these magics strings.

NicholasCWK committed:
Add assert statement

NicholasCWK committed:
Merge pull request #2 from NicholasCWK/branch-A-CodeQuality

Branch a code quality

NicholasCWK committed:
Merge branch 'master' into branch-A-Assertions

* master:
  Refactor Fred, Parser and TaskList class
  Add Action class
  Add Command and TaskType enums

# Conflicts:
#	src/main/java/fred/Fred.java

NicholasCWK committed:
Add assert statement

The parseInput method returns an Action that has a Command attribute.

The executeAction method receives the Action and executes the respective actions based on the value of the Command attribute.

The Command attribute may have an invalid value due to mistakes in the code.

Let’s add an assert statement that checks if the value of the command attribute of the action returned by parseInput is a valid command.

This allows the runtime to detect a possible bug.

NicholasCWK committed:
Merge branch 'branch-A-Assertions'

* branch-A-Assertions:
  Add assert statement
  Add assert statement

NicholasCWK committed:
Add tagging tasks functionality

NicholasCWK committed:
Add EmptyKeywordsException

NicholasCWK committed:
Replace EmptyTagException with InvalidTagException

NicholasCWK committed:
Refactor code in Parser and TaskList to shorten method length

NicholasCWK committed:
Fred.java: Change indentation in switch statement
Task.java: Modify getTags method

NicholasCWK committed:
Remove print statements and unused imports

NicholasCWK committed:
Add InvalidEventException

NicholasCWK committed:
Add InvalidEventException

NicholasCWK committed:
Add InvalidTagException

NicholasCWK committed:
Add InvalidTagException

NicholasCWK committed:
Add isRunning attribute

NicholasCWK committed:
Modify handleUserInput method

NicholasCWK committed:
Modify getTasksFromDataFile method

NicholasCWK committed:
Change the format of by attribute

NicholasCWK committed:
Change the format of from and to attributes

NicholasCWK committed:
Change the format of by attribute

NicholasCWK committed:
Shorten createTask method by abstracting creation of tasks

NicholasCWK committed:
Shorten the executeAction by abstraction

NicholasCWK committed:
Rearrange import statements

NicholasCWK committed:
Add access modifiers

NicholasCWK committed:
Update README.md

NicholasCWK committed:
Update README.md

NicholasCWK committed:
Shorten getTasksFromDataFile method

NicholasCWK committed:
Modify createTask method to catch invalid format

NicholasCWK committed:
Rename all instances of Duke to Fred

NicholasCWK committed:
Add javadocs

NicholasCWK committed:
Update Junit tests

NicholasCWK committed:
Add Ui.png

NicholasCWK committed:
Merge branch 'master' of https://github.com/NicholasCWK/ip

* 'master' of https://github.com/NicholasCWK/ip:
  Update README.md
  Update README.md

NicholasCWK committed:
Update README.md

NicholasCWK committed:
Change access modifiers

NicholasCWK committed:
Add javadocs to exceptions

NicholasCWK committed:
Merge branch 'master' of https://github.com/NicholasCWK/ip

* 'master' of https://github.com/NicholasCWK/ip:
  Update README.md
